Feature/fallback fetch proposals

atharen opened the pull request.

atharen committed:
feat(govStore): fallback for fetching proposals

- Updated fetchProposals method to utilize cache and handle fallback logic for sequential proposal fetching.
- Added caching mechanism for government proposals to improve performance and reduce redundant API calls.

atharen committed:
fixup! add pagination

atharen committed:
fixup! filter spam

atharen committed:
fixup! fix filter issue

atharen committed:
fixup! feat(govStore): fallback for fetching proposals

atharen committed:
refactor(gov): fetch proposals on component mount

- Proposal fetching using Promise.all to reduce redundant API calls.

atharen committed:
refactor(gov): split governance proposal caching

- caching moved to govCache for readability

atharen committed:
feat(gov): caching request intervals

- Introduced separate TTL constants for data and request caching.
- isExpired method to check cached request validity
- manages proposal fetching based on freshness.

migueldingli1997 commented on src/modules/[chain]/gov/index.vue:
Is this change necessary? 🫡 I'd rather not touch anything that is not

atharen replied:
not necessary now - reverted in 33d0e95c13a9f0a1150b844b7949c981f737e41f

atharen committed:
Revert "refactor(gov): fetch proposals on component mount"

This reverts commit 709510edbfe105cea0561d6012dec3b509e83891.

migueldingli1997 commented on src/stores/useGovStore.ts:
```diff
@@ -29,20 +37,89 @@ export const useGovStore = defineStore('govStore', {
+      const cached = GovProposalCache.get(this.blockchain.chainName, status);
+
+      let proposals;
+      let lastKnownProposalId = 0;
```
Consider storing this as a part of the cache, so you don't have to run `Math.max(...cached.proposals.map(p => parseInt(p.proposal_id)));` each time.

atharen committed:
refactor(gov): fallback logic into govFallback

- easier to distinguish with upstream

atharen committed:
Revert "fixup! add pagination"

This reverts commit 38694fb820e6b936cddfd03c69df632ff7755cc6.

atharen commented on src/stores/govCache.ts:
```diff
@@ -0,0 +1,119 @@
+
+export class GovProposalCache {
+    // TTL for data; how long to keep the data in cache
+    private static readonly DATA_CACHE_TTL = 24 * 60 * 60 * 1000; // 1 day cache
```
- 1 minute for voting/2

- unlimited for other status'es

atharen replied:
fixed in 7ee3765e6025c28ebed7621cc2901fc91050b90b

atharen committed:
fixup! refactor(gov): fallback logic into govFallback

atharen committed:
refactor(gov): update caching logic for proposal statuses

- Introduced a new TTL for non-complete proposals, allowing for a 1-minute cache duration.
- Added a method to determine TTL based on proposal status, ensuring completed proposals (status 3 or 4) never expire from the cache.

atharen committed:
fixup! refactor(gov): update caching logic for proposal statuses

atharen committed:
feat(gov): add fallback data notice

- Introduced a fallback data banner in the governance component to inform users when fallback data is being used.
- Updated internationalization files to include translations for the fallback data notice in multiple languages.
- Enhanced the fallback logic to return a flag indicating whether fallback data is in use.

atharen committed:
fixup! feat(gov): add fallback data notice
